ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll

```diff
@@ -13,7 +13,18 @@ class TypeTrackingNode = DataFlowPublic::LocalSourceNode;
 
 predicate simpleLocalFlowStep = DataFlowPrivate::localFlowStepTypeTracker/2;
 
-predicate jumpStep = DataFlowPrivate::jumpStep/2;
+predicate jumpStep(Node nodeFrom, TypeTrackingNode nodeTo) {
```

Comment:
I don't think this should be a jump step, it should be local.

Please correct me if I'm wrong, but as I understand it, a "jump" step is one that resets the calling context (allowing flow out of a different caller than where it came in). Any reason to allow that here?

Reply:
I just (ab)used the `jumpStep` predicate because that has a `LevelStep` summary in `smallstepNoCall`. Perhaps it would be better to introduce a new predicate (`localStep` ?) in the shared TypeTracking library.

Reply:
Actually, it should perhaps be part of `localFlowStepTypeTracker` instead, since then it will work recursively (i.e., a method that returns one of its arguments via a call to another method that also does). This means such summarized steps will also be included in `LocalSourceNode::flowsTo`, which I guess we want?

Note that if we do this, we need to add a new disjunct to `isLocalSourceNode` which includes all nodes that correspond to a `ExprNodes::CallCfgNode`.

Reply:
@hvitved  I added the new step to `simpleLocalFlowStep`. Somehow this changes the expected output back to the previous version. Not sure if that is desirable or not. I am not sure if/how this can be added to `localFlowStepTypeTracker` instead. I ran into some non-monotonic recursion errors.

Reply:
I have added a commit that adds summarized local steps to `LocalSourceNode::flowsTo`, which in turn means that summarized steps will be constructed recursively. @asgerf : Does JS do this as well?

Reply:
No we don't contribute these to the `flowsTo` relation. It would make `flowsTo` depend on the call graph which is a pretty bad thing to do IMO.

In JS the dependency layers here roughly are:
- Local flow, including `flowsTo`
- Import resolution (establish flow between export and import). There are many module systems in JS, and their models may depend on local flow. Other languages with a less chaotic history of module systems may want to put this stage higher up.
- Call resolution and type-tracking steps (mutually recursive)
- API graphs

Reply:
When I said "local" earlier I guess I should have said "level".

Reply:
OK, I have given this some more thought, and I now understand why @aibaars used `jumpStep`, namely because it uses `LevelStep`. So unless we change the interface to the shared type tracking library, I think that is the best we can do for now. I will revert my changes and go back to using `jumpStep`.

Reply:
With https://github.com/github/codeql/pull/8317 we should change it to using `levelStep` instead.

Reply:
@hvitved Shall we close this pull request, or rebase it onto your work in https://github.com/github/codeql/pull/8317 ?

Reply:
Let's rebase once that PR is merged.

Reply:
@aibaars : That PR has now been merged, so let's rebase this PR and use the newly introduced `levelStep`.